feat: add fortune telling app (#62167)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-fortune-teller-app/68e420817107eb7cb2d4d4d6.md

---
id: 68e420817107eb7cb2d4d4d6
title: Step 14
challengeType: 0
dashedName: step-14
---

# --description--

> Add an `onerror` handler to the image element to handle loading failures. On the first failure, mark it using `dataset.failed` and replace the source with a local default image. If the image fails again, remove the hidden class and hide the loading placeholder using `previousElementSibling`.

Reviewer: ```suggestion
Add an `onerror` handler to the image element to handle loading failures. On the first failure, mark it using `dataset.failed` and replace the source with the local default image. If the image fails again, remove the hidden class and hide the loading placeholder using `previousElementSibling`.
```

# --hints--

> The onerror event must be attached to the `img` element.

Reviewer: ```suggestion
You should have an `onerror` handler attached to the `img` element.
```

```js
const onErrorRegex = /onerror\s*=\s*["'][^"']*["']/;
assert.match(code, onErrorRegex)
```

> An `if` statement must be placed inside the `onerror` handler.

Reviewer: ```suggestion
You should have an `if` statement inside the `onerror` handler.
```

```js
const ifInOnErrorRegex = /onerror\s*=\s*["'][^"']*if\s*\(/;
assert.match(code, ifInOnErrorRegex)
```

The `if` condition must check whether `dataset.failed` is `false`.

```js
const datasetCheckRegex = /if\s*\(\s*!this\.dataset\.failed\s*\)/;
assert.match(code, datasetCheckRegex)
```

Inside the if block, `dataset.failed` must be updated to `1`.

```js
const setFailedRegex = /this\.dataset\.failed\s*=\s*['"]1['"]/;
assert.match(code, setFailedRegex)
```

The image `URL` must then be updated to `LOCAL_DEFAULT_IMG`.

```js
const setSrcRegex = /this\.src\s*=\s*['"]\$\{LOCAL_DEFAULT_IMG\}['"]|this\.src\s*=\s*LOCAL_DEFAULT_IMG/;
assert.match(code, setSrcRegex)
```

> The else block must remove the `hidden` class.

Reviewer: ```suggestion
Your else block must remove the `hidden` class.
```
